Default simulation configuration parser

romancardenas opened the pull request:
This first version of the `build_simulation_scenario` function is able to check that the common and default simulation configuration parameters are correct.

The next step is detecting all the entities (i.e., cells) and add them to the output file. To do so, I'll contact @staubibr so I get an idea of how to detect these cells from existing structures.

francoiseblanc commented on test/test_simulation.c:
```diff
@@ -0,0 +1,90 @@
+    read_json_file("./data/simulation/2_3_nonexisting_model.json", &json);
+    TEST_ASSERT_EQUAL(SIM_MODEL_SELECTION_INVALID, build_simulation_scenario(json, NULL, NULL));
+}
+
```
Only your first test case is executed. I believe that you have to change all your paths to include: ../test/data/simulation

romancardenas replied:
Which is your working directory? I selected the `test` folder as the working directory when running the tests. Nonetheless, I'll check the paths as you suggested.

francoiseblanc commented on test/test_simulation.c:
```diff
@@ -0,0 +1,90 @@
+    read_json_file("./data/simulation/5_2_invalid_output_config_path.json", &json);
+    TEST_ASSERT_EQUAL(SIM_CONFIG_OUTPUT_PATH_INVALID, build_simulation_scenario(json, NULL, NULL));
+
+    read_json_file("./data/simulation/5_3_existing_output_config_path.json", &json);
```
I think the path in this .json file must also be changed. Or the error code must be changed.

francoiseblanc reviewed (approved):
Great work! Just a few comments regarding test_simulation.c

romancardenas commented:
> Great work! Just a few comments regarding test_simulation.c

Thanks, @francoiseblanc ! I've just pushed a new commit addressing your comments. Please, confirm that now the tests run successfully

romancardenas commented on src/simulation_lifecycle/models.c:
```diff
@@ -0,0 +1,170 @@
+#define DELAY_BUFFER_HYBRID "hybrid"
+#define DELAY_BUFFER_INERTIAL "inertial"
+
+#define MODEL_SIR "sir"
+#define MODEL_SIRS "sirs"
+#define MODEL_SIRD "sird"
+#define MODEL_SIRDS "sirds"
+
+#define MODEL_SIRX_POPULATION "population"
+#define MODEL_SIRX_SUSCEPTIBLE "susceptible"
+#define MODEL_SIRX_INFECTED "infected"
+#define MODEL_SIRX_RECOVERED "recovered"
+#define MODEL_SIRX_DECEASED "deceased"
+
+#define MODEL_SIRX_VIRULENCE "virulence"
+#define MODEL_SIRX_RECOVERY "recovery"
+#define MODEL_SIRX_FATALITY "fatality"
+#define MODEL_SIRX_IMMUNITY "immunity"
```
I use these macros to check that the specified model in the workflow exists. Also, I check that all the required configuration parameters exist and are of the expected data type (in this case, all are floats). We can remove this and make it less strict. But then simulation may fail.

staubibr replied:
Well you could have the info you need in the simulation section. In addition to the mapping, we would have to add the field type.

I dont mind though. If it adds too much complexity, its not worth it.

romancardenas commented on src/simulation_lifecycle/models.c:
```diff
@@ -0,0 +1,170 @@
+
+    /* 2. Copy configuration parameters (checking that all the required parameters are there) */
+    cJSON *default_config = cJSON_CreateObject();
+    cJSON *c = cJSON_GetObjectItemCaseSensitive(from, MODEL_CELL_CONFIG);
+    if (c == NULL || !cJSON_IsObject(c) || copy_json_values(cJSON_IsNumber, c, default_config,
+                                                            MODEL_SIRX_VIRULENCE, MODEL_SIRX_RECOVERY,
+                                                            MODEL_SIRX_FATALITY, NULL)) {
+        return SIM_MODEL_SPECIFIC_CONFIG_INVALID;
+    }
+    cJSON_AddItemToObject(target, MODEL_CELL_CONFIG, default_state);
+
+    return SUCCESS;
+}
+
+int parse_default_sirds_model(const cJSON *from, cJSON *target) {
+    cJSON_AddStringToObject(target, MODEL_CELL_TYPE, "sirds");
+
+    /* 1. Copy state parameters (checking that all the required parameters are there) */
+    cJSON *default_state = cJSON_CreateObject();
+    cJSON *s = cJSON_GetObjectItemCaseSensitive(from, MODEL_CELL_STATE);
+    if (s == NULL || !cJSON_IsObject(s) || copy_json_values(cJSON_IsNumber, s, default_state,
+                                                            MODEL_SIRX_POPULATION, MODEL_SIRX_SUSCEPTIBLE,
+                                                            MODEL_SIRX_INFECTED, MODEL_SIRX_RECOVERED,
+                                                            MODEL_SIRX_DECEASED, NULL)) {
+        return SIM_MODEL_SPECIFIC_CONFIG_INVALID;
+    }
+    cJSON_AddItemToObject(target, MODEL_CELL_STATE, default_state);
+
+    /* 2. Copy configuration parameters (checking that all the required parameters are there) */
+    cJSON *default_config = cJSON_CreateObject();
+    cJSON *c = cJSON_GetObjectItemCaseSensitive(from, MODEL_CELL_CONFIG);
+    if (c == NULL || !cJSON_IsObject(c) || copy_json_values(cJSON_IsNumber, c, default_config,
+                                                            MODEL_SIRX_VIRULENCE, MODEL_SIRX_RECOVERY,
+                                                            MODEL_SIRX_FATALITY, MODEL_SIRX_IMMUNITY, NULL)) {
+        return SIM_MODEL_SPECIFIC_CONFIG_INVALID;
+    }
+    cJSON_AddItemToObject(target, MODEL_CELL_CONFIG, default_state);
+
+    return SUCCESS;
+}
```
If we do a more generic approach, then the function won't be able to check that the simulation configuration is correct. Thus, the simulation may fail. I'm OK with it, we can discuss this in the next meeting and decide whether to change it.

romancardenas commented on src/simulation_lifecycle/utils/file.c:
```diff
@@ -18,4 +21,38 @@ int write_data_to_file(char *filepath, char *data) {
\ No newline at end of file
+}
+
+void read_json_file(char *file_path, cJSON **p_target) {
+    FILE *f = fopen(file_path, "rb");
+    fseek(f, 0, SEEK_END);
+    long fsize = ftell(f);
+    fseek(f, 0, SEEK_SET);  /* same as rewind(f); */
+
+    char *string = malloc(fsize + 1);
+    fread(string, 1, fsize, f);
+    fclose(f);
+    string[fsize] = '\0';
+
+    *p_target = cJSON_Parse(string);
+    free(string);
+}
```
We can merge my branch and when we merge yours, I'll take care of all the collisions (I'm used to working with this Git stuff)

staubibr replied:
I don't mind merging at some point either, I've done it in the past. But I certainly won't say no if you want to do it! I have  an ever growing list of things to do this week...

romancardenas commented:
@staubibr @francoiseblanc 

I significantly changed the code to address Bruno's comments. Now it is more general. It does not need a specific implementation for each model.
- It checks if there is a binary with the name of the selected model
- It checks that important fields (like `state`) exist, but it cannot tell if it contains valid values. Therefore, that issue will arise when running the simulation.

I think it is a more flexible approach, and even though. What do you think?

staubibr reviewed (approved):
Looks good to me, well done!

francoiseblanc commented:
Agreed! Really well done!

Is it expected that two of your tests are failing?

 
![image](https://user-images.githubusercontent.com/71035708/100393115-96085e80-3006-11eb-8c65-1cf1efec68f2.png)
